### PR TITLE
Analyze/show more query types

### PR DESCRIPTION
**Pull request by DL6ER**
**By submitting this pull request, I confirm the following (please check boxes, eg [X]) _Failure to fill the template will close your PR_:**

***Please submit all pull requests against the `development` branch. Failure to do so will delay or deny your request***

- [X] I have read and understood the [contributors guide](https://github.com/pi-hole/pi-hole/blob/master/CONTRIBUTING.md).
- [X] I have checked that [another pull request](https://github.com/pi-hole/FTL/pulls) for this purpose does not exist.
- [X] I have considered, and confirmed that this submission will be valuable to others.
- [X] I accept that this submission may not be used, and the pull request closed at the will of the maintainer.
- [X] I give this submission freely, and claim no ownership to its content.

**How familiar are you with the codebase?:**

## 10

---

So far, Pi-hole analyzed only `A` and `AAAA` queries. We open *FTL*DNS now for many additional query types, namely `PTR`, `SRV`, `SOA`, `TXT`, and `ANY`.

This feature is theoretically ready but intentionally marked as `WIP` as it needs more testing and reviews by all volunteers.

Screen shot:
screenshot at 2018-08-09 17-59-45

_This template was created based on the work of [`udemy-dl`](https://github.com/nishad/udemy-dl/blob/master/LICENSE)._

### CODE REVIEW

**Comment by Mcat12** [comment]
Note: This PR also fixes the over 100% block rate bug. 1f360dc

File: dnsmasq_interface.c [hunk 1]
```diff
@@ -396,96 +393,72 @@ void FTL_reply(unsigned short flags, char *name, struct all_addr *addr, int id)
...
-		if(!queries[i].complete)
-		{
-			// This query is no longer unknown
-			counters.unknown--;
-			// Answered from a custom (user provided) cache file
-			counters.cached++;
+	if(!found)
+	{
+		// This may happen e.g. if the original query was a PTR query or "pi.hole"
+		// as we ignore them altogether
```
Comment (Mcat12) [inline dnsmasq_interface.c]: This comment is now out of date

File: dnsmasq_interface.c [hunk 2]
```diff
@@ -396,96 +393,72 @@ void FTL_reply(unsigned short flags, char *name, struct all_addr *addr, int id)
...
+		return;
+	}
 
-			if(strcmp(answer, "(NXDOMAIN)") == 0 ||
-			   strcmp(answer, "0.0.0.0") == 0 ||
-			   strcmp(answer, "::") == 0)
-				queries[i].status = QUERY_WILDCARD;
-			else
-				queries[i].status = QUERY_CACHE;
+	if(queries[i].reply != REPLY_UNKNOWN)
```
Comment (Mcat12) [inline dnsmasq_interface.c]: Would checking `queries[i].complete` do the same thing? (that is what is used in the old code)

File: dnsmasq_interface.c [hunk 3]
```diff
@@ -396,96 +393,72 @@ void FTL_reply(unsigned short flags, char *name, struct all_addr *addr, int id)
...
-			validate_access("overTime", timeidx, true, __LINE__, __FUNCTION__, __FILE__);
+	if(flags & F_CONFIG)
+	{
+		// Answered from local configuration, might be a wildcard or user-provided
+		// This query is no longer unknown
+		counters.unknown--;
+		// Answered from a custom (user provided) cache file
+		counters.cached++;
+
+		// Detect user-defined blocking rules in .conf files
```
Comment (Mcat12) [inline dnsmasq_interface.c]: What does "in .conf files" reference? The comment still makes sense without it.

File: dnsmasq_interface.c (hunk 2 above)
Comment (DL6ER) [reply dnsmasq_interface.c]: No, not in the following case:

1. FTL receives a query (FTL creates `queries[i]`)
2. `dnsmasq` decides to forward it (FTL sets `queries[i].status = QUERY_FORWARDED` and queries[i].complete = true` as the query status is now known)
3. When we get the reply from upstream we have to check if the `reply` is already known (this is this code)

**Comment by pralor** [comment]
This pull request has been mentioned on **Pi-hole Userspace**. There might be relevant details there:

https://discourse.pi-hole.net/t/where-to-find-queries-by-type-ptr/12528/2

**Comment by walter-exit** [comment]
Good evening lads and lasses,

I see we all noticed this odd behavior. And the amazing folks at Pi-hole seem to have found and fixed the culprit. Perfect! Could this now be released?

The current release still contains this annoying bug. Two agonizing months, since I discovered it, looking at askew percentages bugs me more than I like to admit.

Can we save Christmas this year and release the development fixes so far? Or perhaps just a hot fix for this one? That would sure make my holiday season ;-)

Happy days to you all, in advance.

**Comment by pralor** [comment]
This pull request has been mentioned on **Pi-hole Userspace**. There might be relevant details there:

https://discourse.pi-hole.net/t/never-forward-reverse-lookups-for-private-ip-ranges-doesnt-appear-to-be-working-in-4-1-1/15614/2

**Pull request merged**